src/FilenameParsing/AbstractFileIDHelper.php

```diff
@@ -0,0 +1,153 @@
+     */
+    public const EXTENSION_REWRITE_VARIANT = 'extRewrite';
+
+    /**
+     * use the original file's extension
+     */
+    private const EXT_ORIGINAL = 0;
+
+    /**
+     * use the variant file's extension
+     */
+    private const EXT_VARIANT = 1;
+
+    public function buildFileID($filename, $hash = null, $variant = null, $cleanfilename = true)
+    {
+        if ($filename instanceof ParsedFileID) {
+            $hash =  $filename->getHash();
+            $variant =  $filename->getVariant();
+            $filename =  $filename->getFilename();
+        }
+
+        $this->validateFileParts($filename, $hash, $variant);
+
+        // Since we use double underscore to delimit variants, eradicate them from filename
+        if ($cleanfilename) {
+            $filename = $this->cleanFilename($filename);
+        }
+
+        if ($variant) {
+            $filename = $this->rewriteVariantExtension($filename, $variant);
+        }
+
+        $name = basename($filename ?? '');
+
+        // Split extension
+        $extension = null;
+        if (($pos = strpos($name ?? '', '.')) !== false) {
+            $extension = substr($name ?? '', $pos ?? 0);
+            $name = substr($name ?? '', 0, $pos);
+        }
+
+        $fileID = $this->getFileIDBase($name, $filename, $hash, $variant);
+
+        // Add directory
+        $dirname = ltrim(dirname($filename ?? ''), '.');
+        if ($dirname) {
+            $fileID = $dirname . '/' . $fileID;
+        }
+
+        // Add variant
+        if ($variant) {
+            $fileID .= '__' . $variant;
+        }
+
+        // Add extension
+        if ($extension) {
+            $fileID .= $extension;
+        }
+
+        return $fileID;
+    }
+
+    /**
+     * Get the original filename with the extension rewritten to be the same as the variant file extension.
+     *
+     * @param string $filename Original filename without variant
+     */
+    public function rewriteVariantExtension(string $filename, string $variant): string
+    {
+        return $this->swapExtension($filename, $variant, self::EXT_VARIANT);
+    }
+
+    /**
+     * Get the variant filename with the extension rewritten to be the same as the original file extension.
+     *
+     * @param string $filename Original filename without variant
+     */
+    public function restoreOriginalExtension(string $filename, string $variant): string
+    {
+        return $this->swapExtension($filename, $variant, self::EXT_ORIGINAL);
+    }
+
+    /**
+     * Get the original file's filename with the extension rewritten to be the same as either the original
+     * or the variant extension.
+     *
+     * @param string $filename Original filename without variant
+     * @param int $extIndex One of self::EXT_ORIGINAL or self::EXT_VARIANT
+     */
+    private function swapExtension(string $filename, string $variant, int $extIndex): string
```

Comment:
Swapping the extension is necessary so we can correctly identify the file on the filesystem. Without this, the original file extension is used when searching for the variant file in the filesystem.

e.g. you have an original file `myfile.mp4`, and a variant `myfile__extRewriteWyJtcDQiLCJqcGciXQ.jpg`.
Without this change, if would look for `myfile__extRewriteWyJtcDQiLCJqcGciXQ.mp4` on the filesystem, and obviously it wouldn't find it.